org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/codemanipulation/DefaultJavaElementComparator.java

```diff
@@ -0,0 +1,151 @@
+			case ASTNode.METHOD_DECLARATION:
+				{
+					MethodDeclaration method= (MethodDeclaration) bodyDeclaration;
+					if (method.isConstructor()) {
+						return MembersOrderPreferenceCacheCommon.CONSTRUCTORS_INDEX;
+					}
+					int flags= method.getModifiers();
+					if (Modifier.isStatic(flags))
+						return MembersOrderPreferenceCacheCommon.STATIC_METHODS_INDEX;
+					else
+						return MembersOrderPreferenceCacheCommon.METHOD_INDEX;
+				}
+			case ASTNode.FIELD_DECLARATION :
+				{
+					if (JdtFlags.isStatic(bodyDeclaration))
+						return MembersOrderPreferenceCacheCommon.STATIC_FIELDS_INDEX;
+					else
+						return MembersOrderPreferenceCacheCommon.FIELDS_INDEX;
+				}
+			case ASTNode.INITIALIZER :
+				{
+					int flags= ((Initializer) bodyDeclaration).getModifiers();
+					if (Modifier.isStatic(flags))
+						return MembersOrderPreferenceCacheCommon.STATIC_INIT_INDEX;
+					else
+						return MembersOrderPreferenceCacheCommon.INIT_INDEX;
+				}
+			case ASTNode.TYPE_DECLARATION :
+			case ASTNode.ENUM_DECLARATION :
+			case ASTNode.ANNOTATION_TYPE_DECLARATION :
+				return MembersOrderPreferenceCacheCommon.TYPE_INDEX;
+			case ASTNode.ENUM_CONSTANT_DECLARATION :
+				return MembersOrderPreferenceCacheCommon.ENUM_CONSTANTS_INDEX;
+			case ASTNode.ANNOTATION_TYPE_MEMBER_DECLARATION:
+				return MembersOrderPreferenceCacheCommon.METHOD_INDEX; // reusing the method index
+
+		}
+		return 0; // should never happen
+	}
+
+	private int getCategoryIndex(int category) {
+		return fMemberOrderCache.getCategoryIndex(category);
+	}
+
+	/**
+	 * This comparator follows the contract defined in CompilationUnitSorter.sort.
+	 * @see Comparator#compare(java.lang.Object, java.lang.Object)
+	 * @see CompilationUnitSorter#sort(int, org.eclipse.jdt.core.ICompilationUnit, int[], java.util.Comparator, int, org.eclipse.core.runtime.IProgressMonitor)
+	 */
+	@Override
+	public int compare(BodyDeclaration bodyDeclaration1, BodyDeclaration bodyDeclaration2) {
+		boolean preserved1= fDoNotSortFields && isSortPreserved(bodyDeclaration1);
+		boolean preserved2= fDoNotSortFields && isSortPreserved(bodyDeclaration2);
+
+		// Bug 407759: need to use a common category for all isSortPreserved members that are to be sorted in the same group:
+		int cat1= category(bodyDeclaration1);
+		if (preserved1) {
+			cat1= sortPreservedCategory(cat1);
+		}
+		int cat2= category(bodyDeclaration2);
+		if (preserved2) {
+			cat2= sortPreservedCategory(cat2);
+		}
+
+		if (cat1 != cat2) {
+			return getCategoryIndex(cat1) - getCategoryIndex(cat2);
+		}
+
+		// cat1 == cat2 implies preserved1 == preserved2
+
+		if (preserved1) {
+			return preserveRelativeOrder(bodyDeclaration1, bodyDeclaration2);
+		}
+
+		int flags1= JdtFlags.getVisibilityCode(bodyDeclaration1);
+		int flags2= JdtFlags.getVisibilityCode(bodyDeclaration2);
+		int vis= fMemberOrderCache.getVisibilityIndex(flags1) - fMemberOrderCache.getVisibilityIndex(flags2);
+		return vis;
```

Comment:
> So just to confirm, the following block is left out to avoid alphabetical sorting

yes.

My first thought in this proposal is that we can just remove the alphabetical sorting, and if there is any request about this, we can consider enabling it via a configuration (also add these logics back). In jdt.ui implementation, the alphabetical sorting is enabled and seems can't be disabled, so I simply remove them, including the `Collator` since it's only used in the alphabetical sorting block.

Do you think we should keep it and make it configurable now?

Reply:
Right. So since there's only options for configuring the sort categories, and visibility, it makes sense to only sort by those since users have no control over the name sorting.

What about examples like :

```
private String someFieldA;
public String someFieldB = someFieldA;
```

Sorting would generate a compilation error here (visibility sorting). In fact, this example only further confirms sorting by name may not be such a great idea as it might be even more likely to produce some bad behaviour. Should we set `fDoNotSortFields` to `true` on our end or are we fine with this consequence ? Note that in Eclipse, `fDoNotSortFields` is used to switch between sorting all elements, or not sorting fields, initializers, enums :

![sort_members_options](https://user-images.githubusercontent.com/1417342/182418432-a1f221d1-67d6-48cb-b865-04a460295d8f.png)

Reply:
> Right. So since there's only options for configuring the sort categories, and visibility, it makes sense to only sort by those since users have no control over the name sorting.

1. Keep current PR behavior. (disable alphabetical sorting and leave alphabetical sorting part out of the file).
2. Keep alphabetical sorting part in the file, make it configurable via a VS Code configuration, as for the default value I prefer **disable** the alphabetical sorting.

@rgrunber which one do you prefer?

As for the `fDoNotSortFields` part, I agree that we can disable it by default to avoid any bad behavior. Besides, we can also make it configurable via a VS Code configuration like eclipse does.

Reply:
Let's keep the alphabetical sorting out for now, **(1)**, and we can re-evaluate at another time.